### PR TITLE
Set defaults on correlation id and causation id

### PR DESCRIPTION
**Pull request by nlunn**
This is the first version of the change. 
I still need to be able to handle outbox
#40

### CODE REVIEW

File: src/Dafda/Serializing/PayloadDescriptor.cs [hunk 1]
```diff
@@ -25,8 +27,10 @@ public PayloadDescriptor(string messageId, string topicName, string partitionKey
             PartitionKey = partitionKey;
             MessageType = messageType;
             MessageData = messageData;
-            MessageHeaders = messageHeaders;
-        }
+            MessageHeaders = messageHeaders
+                .Where(k => !string.Equals(k.Key, "messageId", StringComparison.InvariantCultureIgnoreCase))
```
Comment (jakob-nielsen-dfds) [inline src/Dafda/Serializing/PayloadDescriptor.cs:31]: Seems a bit dirty - why do we need to filter out things just passed to the constructor?
Why should `messageId`and `type` not be part of the headers?

Comment (nlunn) [reply src/Dafda/Serializing/PayloadDescriptor.cs:31]: Good spotted :)
if we add headers that are allready mandatory, then the serialization will break. 
I am only safeguarding what is allready broken.

Comment (jakob-nielsen-dfds) [reply src/Dafda/Serializing/PayloadDescriptor.cs:31]: Yeah - dictionary keys needs to be unique :)
Okay - could the knowledge about which fields needs to be filtered out be kept in `DefaultPayloadSerializer` (this seems to define them), maybe in the form of an extension method doing the filtering for you, ie. `FilterOutDefaultHeaders`  or something?

Comment (nlunn) [reply src/Dafda/Serializing/PayloadDescriptor.cs]: Moved to the serializer.

File: src/Dafda.Tests/Producing/TestOutgoingMessageFactory.cs [hunk 2]
```diff
@@ -27,7 +27,7 @@ public async Task Can_create_outgoing_message_from_registry_with_expected_raw_me
 
             await sut.Produce(new DummyMessage(dummyAggregateId));
 
-            Assert.Equal($@"{{""messageId"":""{dummyMessageId}"",""type"":""{DummyType}"",""data"":{{""aggregateId"":""{dummyAggregateId}""}}}}", spy.Value);
+            Assert.Equal($@"{{""messageId"":""{dummyMessageId}"",""type"":""{DummyType}"",""causationId"":""foo_id"",""correlationId"":""foo_id"",""data"":{{""aggregateId"":""{dummyAggregateId}""}}}}", spy.Value);
```
Comment (chamook) [inline src/Dafda.Tests/Producing/TestOutgoingMessageFactory.cs:30]: 📐 This value is getting a bit long and hard to read now, any chance we could break it out to a named value and spread it over a few lines so it's clearer to see what data is in it?

File: src/Dafda/Producing/Producer.cs [hunk 3]
```diff
@@ -53,5 +51,27 @@ public async Task Produce(object message, Dictionary<string, object> headers)
...
+            await Produce(message, context, new Dictionary<string, string>());
+        }
+
+        /// <summary>
+        /// Produce a <paramref name="message"/> on Kafka
+        /// </summary>
+        /// <param name="message">The message</param>
+        /// <param name="context">Context from the consumer</param>
+        /// <param name="headers">The message headers</param>
+        public async Task Produce(object message, MessageHandlerContext context, Dictionary<string, string> headers)
```
Comment (chamook) [inline src/Dafda/Producing/Producer.cs:71]: Hmm the more we expose this as `Dictionary<string, string>` the less opinionated we can (safely) be about the content within - we already have a bit of logic around this with some `string.IsNullOrEmpty` checks, which could be safer to expose to clients through a type where an empty string is not a representable value. Would it be reasonable to expose this as just the `Metadata` type?

**Review by chamook — COMMENTED** [review]
Seems a sensible enough direction to go, just some concerns about how we expose the underlying data (and I might be a bit late to the party on that, sorry 😞 )

File: src/Dafda/Producing/Producer.cs (hunk 3 above)
Comment (nlunn) [reply src/Dafda/Producing/Producer.cs:71]: I do agree. However the problem here is that the context is filling out the CorrelationId and the CausationId. By adding the Metadata object, we have a conflict with which object to take those named fields from.
I just belive that when implemented this way, it is much easier understandable where data for the named fields are coming from :)

File: src/Dafda/Serializing/DefaultPayloadSerializer.cs [hunk 4]
```diff
@@ -49,16 +51,20 @@ private object ConvertToMessagePayload(PayloadDescriptor descriptor)
...
-                    MakeKeyFrom("Type"), 
+                    MakeKeyFrom("Type"),
                     descriptor.MessageType
                 }
             };
 
-            foreach (var (key, value) in descriptor.MessageHeaders)
+            var messageHeaders = descriptor.MessageHeaders
+                .Where(k => !string.Equals(k.Key, "messageId", StringComparison.InvariantCultureIgnoreCase))
+                .Where(k => !string.Equals(k.Key, "type", StringComparison.InvariantCultureIgnoreCase));
```
Comment (jakob-nielsen-dfds) [inline src/Dafda/Serializing/DefaultPayloadSerializer.cs:65]: I like it better to keep the logic in the serializer 👍

But I don't really know why these headers are filtered out from reading (just this part of) the code. Would a named list describing what headers are already defined/default (or better naming), using this to filter out entries, make it more clear?

Something like `descriptor.MessageHeaders.Where(header => !DefaultHeaders.Contains(header));`

Comment (jakob-nielsen-dfds) [reply src/Dafda/Serializing/DefaultPayloadSerializer.cs:65]: Or an extension method filtering default headers out, something like `descriptor.MessageHeaders.ExcludeDefaultHeaders()` or `descriptor.MessageHeaders.WithoutDefaultHeaders()`.

Comment (nlunn) [reply src/Dafda/Serializing/DefaultPayloadSerializer.cs:65]: There may be better ways to implement this. I did not focus more on this than fixing what was broken. 
My main focus for this pull request is differently and I would like to keep that focus. Is that ok for now?

**Review by chamook — APPROVED** [review]
Looks good to me, a nice addition 😄  :+1:

**Pull request merged**